### ISSUE TITLE
min_bin_size/min_prebin_size silently become a floor on class-reweighted mass instead of row count once class_weight is set (behavior differs between 0.19.x and current)

### ISSUE THREAD

**Issue by FrancoRamos** [guillermo-navas-palencia/optbinning#385]
## Environment

- `optbinning` versions compared: `0.19.0` and `0.21.0` (current `pip` release at time of writing)
- Python 3.11, Windows
- `scikit-learn==1.9.0`, `numpy==2.4.6` (used with `optbinning==0.21.0`)
- (`optbinning==0.19.0` was tested in an isolated venv against the same synthetic data below; note it predates `scikit-learn==1.9.0` and needs an older `scikit-learn` to run at all — `check_array(..., force_all_finite=...)` was renamed in newer `scikit-learn`, unrelated to this report)

## Summary

When `class_weight` is set (e.g. `"balanced"`) on a severely imbalanced binary target, `min_bin_size`/`min_prebin_size` stop guaranteeing "each bin holds at least X% of the rows" and instead guarantee "each bin holds at least X% of the *class-reweighted* pseudo-mass". Because `binning_table.build()`'s own `Count`/`Count (%)`/`Event rate` columns are also computed from that reweighted view, the summary table doesn't reveal the discrepancy — a final bin can hold well under 1% of real rows while being reported (and treated internally) as satisfying a 5% floor.

This reproduces on `0.21.0`. The same data on `0.19.0` does not exhibit it — `PreBinning.fit`'s CART branch used to enforce `min_samples_leaf` (an absolute row count) unconditionally; a later version switched to `min_weight_fraction_leaf` whenever `sample_weight` is not `None`, which it always is once `class_weight` triggers the internal `sample_weight` construction in `preprocessing.py::split_data`. I'm not sure if this was an intentional change (see questions below).

## Root cause (as far as I traced it)

`optbinning/binning/prebinning.py::PreBinning.fit`, `method == "cart"` branch:

**0.19.0** (always row-count-based, regardless of `sample_weight`):
```python
cart_kwargs = {
        "min_samples_leaf": self.min_bin_size,
        "max_leaf_nodes": self.n_bins}

if self.problem_type == "classification":
    cart_kwargs["class_weight"] = self.class_weight
    cart_kwargs.update(**self.kwargs)
    est = DecisionTreeClassifier(**cart_kwargs)
...
est.fit(x.reshape(-1, 1), y, sample_weight=sample_weight)
```

**0.21.0** (switches to a weighted-fraction floor whenever `sample_weight` is not `None`):
```python
if sample_weight is None:
    cart_kwargs = {
            "min_samples_leaf": self.min_bin_size,
            "max_leaf_nodes": self.n_bins}
else:
    # https://scikit-learn.org/stable/modules/tree.html#tips-on-practical-use
    # If the samples are weighted, it will be easier to optimize the tree
    # structure using weight-based pre-pruning criterion such as
    # min_weight_fraction_leaf, which ensure that leaf nodes contain at
    # least a fraction of the overall sum of the sample weights.
    cart_kwargs = {
        "min_weight_fraction_leaf": min(
            0.5, self.min_bin_size / np.sum(sample_weight)
            ),
        "max_leaf_nodes": self.n_bins
    }
```

`class_weight` → `sample_weight` conversion itself (`preprocessing.py::split_data`, via `sklearn.utils.compute_class_weight`) is unchanged between both versions, as far as I can tell — the difference is specifically in how `PreBinning.fit` consumes that `sample_weight` for the leaf-size constraint.

## Reproducible example

Synthetic data only (no proprietary data involved) — chosen to mimic a common credit-risk WoE-binning shape: ~98% of rows in a low/homogeneous-risk region, ~2% in a high-risk tail, ~1.6% overall event rate (a fairly typical severe class imbalance in this kind of use case).

```python
import numpy as np
import optbinning

print("optbinning version:", optbinning.__version__)

rng = np.random.default_rng(0)
n = 200_000

n_bulk = int(n * 0.984)
n_tail = n - n_bulk

x_bulk = rng.uniform(0, 60, n_bulk)
x_tail = rng.uniform(60, 90, n_tail)
x = np.concatenate([x_bulk, x_tail])

y_bulk = (rng.random(n_bulk) < 0.0013).astype(int)
y_tail = (rng.random(n_tail) < 0.93).astype(int)
y = np.concatenate([y_bulk, y_tail])

idx = rng.permutation(n)
x, y = x[idx], y[idx]

print(f"n={n}, overall event rate={y.mean():.4%}")

optb = optbinning.OptimalBinning(
    name="x", dtype="numerical", solver="cp",
    monotonic_trend="auto_asc_desc",
    class_weight="balanced",
    min_prebin_size=0.05, min_bin_size=0.05,
)
optb.fit(x, y)

print("status:", optb.status)
print("splits:", optb.splits)
print(optb.binning_table.build()[["Bin", "Count", "Count (%)", "Event", "Event rate"]])

# Re-bin the RAW x with the SAME split points, no reweighting, to show the real row share.
edges = [-np.inf] + list(optb.splits) + [np.inf]
for lo, hi in zip(edges[:-1], edges[1:]):
    mask = (x >= lo) & (x < hi)
    print(f"[{lo}, {hi})  raw row share = {mask.mean()*100:.3f}%")
```

### Actual output on `optbinning==0.21.0`

```
optbinning version: 0.21.0
n=200000, overall event rate=1.6200%
status: OPTIMAL
splits: [59.77811813 65.42506027]

           Bin  Count  Count (%)  Event  Event rate
 (-inf, 59.78) 107597   0.538015   8086    0.075151
[59.78, 65.43)  15977   0.079889  15585    0.975465
  [65.43, inf)  76415   0.382096  76323    0.998796
       Special      0   0.000000      0    0.000000
       Missing      0   0.000000      0    0.000000
        Totals 199989   1.000000  99994    0.499997
```

Real row share per bin, applying the exact same split points to the raw (unweighted) `x`:

```
[-inf, 59.778)   raw row share = 98.031%
[59.778, 65.425) raw row share =  0.639%
[65.425, inf)    raw row share =  1.331%
```

So `optbinning` reports (and internally treated as passing `min_bin_size=0.05`) three bins of `53.8% / 8.0% / 38.2%` — all comfortably over 5% — while the true row-count split is `98.0% / 0.6% / 1.3%`, two of which are far under the 5% floor the parameters are supposed to guarantee.

On `optbinning==0.19.0`, the same script (same seed, same data) collapses to a single bin (`splits: []`) instead — i.e. it correctly refuses to create a bin under the 5% real-row floor, since the minority-risk tail here is only ~2% of rows total.

## Why this matters

The `>=X%` per bin convention in WoE/credit-scorecard binning is meant as a bin-stability guarantee measured in real accounts/observations — a bin backed by very few real rows has an unstable WoE estimate that can swing a lot between refreshes, which is exactly the risk `min_bin_size`/`min_prebin_size` are documented to protect against. Under `0.21.0`, that guarantee silently stops applying once `class_weight` is used with an imbalanced target — which is precisely the situation where `class_weight="balanced"` is usually needed in the first place (without it, CART often can't find any split at all on this kind of imbalance; I confirmed that too, it collapses to one bin regardless of version).

## Questions

1. Is the switch to `min_weight_fraction_leaf` when `sample_weight is not None` an intentional design decision (e.g. following scikit-learn's own recommendation for weighted trees), or a side effect of a broader refactor?
2. If intentional: would it be reasonable to document this explicitly on `min_bin_size`/`min_prebin_size` (and/or on `class_weight`) — i.e. make clear that once `class_weight`/`sample_weight` is used, "X%" refers to reweighted mass, not row count?
3. Would you be open to an opt-in flag to keep the old (row-count) semantics available for callers who want it, for cases like WoE/scorecard binning where the row-count reading is the intended one? I put together a small local workaround (overriding just this branch of `PreBinning.fit`) to keep using `0.21.0` while getting the old behavior back on demand — happy to share details if useful, wasn't sure if you'd want that included here as a suggested diff or would rather decide the right shape for a fix yourselves.

## Related

- #64 ("Sample Weights Error") touches this same `sample_weight`/prebinning code path.
- #357 ("Error when no split founds in ContinuousOptimalBinning") is about the degenerate-single-bin case, which is related but distinct from this report.

**Comment by lcrmorin** [comment]
Seems like a semantics issue. Currently the constraints apply to the weighted population.  It should at least be documented properly. 

However, the real solution seems to be to have two constraints available in the API, one weighted / one unweighted. It might be a tricky to have both at the same time.
